modules/packages/UnrolledLinkedList.chpl

```diff
@@ -0,0 +1,1249 @@
+      _commonInitFromIterable(other);
+    }
+
+    /*
+      Initializes a unrolledLinkedList containing elements that are copy initialized from
+      the elements contained in another unrolledLinkedlist.
+
+      :arg other: The list to initialize from.
+    */
+    proc init=(other: unrolledLinkedList(this.type.eltType, ?p)) {
+      if !isCopyableType(this.type.eltType) then
+        compilerError("Cannot construct unrolledLinkedList from list with " +
+        "element type that cannot be copied");
+      _checkType(this.type.eltType);
+      _checkNodeCapacity(other.nodeCapacity);
+
+      this.eltType = this.type.eltType;
+      this.parSafe = this.type.parSafe;
+      this.nodeCapacity = other.nodeCapacity;
+      this.complete();
+      _commonInitFromIterable(other);
+    }
+
+    pragma "no doc"
+    proc _commonInitFromIterable(iterable) {
+      for x in iterable do
+        append(x);
+    }
+
+    // A helper function for getting a reference to a unrolledLinkedList element.
+    //
+    pragma "no doc"
+    inline proc const ref _getRef(idx: int) ref {
+      _sanity(idx >= 0 && idx < _size);
+      _sanity(_head != nil);
+      var cur = _head, pos = 0;
+      while cur != nil {
+        if idx < pos + cur!.size {
+          ref result = cur!.data[idx - pos];
+          return result;
+        }
+        pos += cur!.size;
+        cur = cur!.next;
+      }
+      halt("unrolledLinkedList out of range");
+    }
+
+    pragma "no doc"
+    inline proc _enter() {
+      if parSafe then
+        _lock$.lock();
+    }
+
+    pragma "no doc"
+    inline proc _leave() {
+      if parSafe then
+        _lock$.unlock();
+    }
+
+    pragma "no doc"
+    inline proc const _withinBounds(idx: int): bool {
+      return (idx >= 0 && idx < _size);
+    }
+
+    /*
+     Split a node from the given node p
+     by moving half of the content of p into the new one
+    */
+    pragma "no doc"
+    proc _split(p: unmanaged _linkedNode(eltType)) {
+      _sanity(p.size == nodeCapacity);
+      var node = new unmanaged _linkedNode(eltType, nodeCapacity);
+      var start = p.size/2;
+      for i in start..p.size-1 {
+        node.append(p.data[i]);
+      }
+      p.size -= node.size;
+      _sanity(p.size > 0);
+
+      // link the new node to the old node
+      node.next = p.next;
+      p.next = node;
+
+      if p == _tail then _tail = p.next;
+
+      return node;
+    }
+
+    /*
+      If it's possible to merge this node and the next node, then do it
+      If it's not possible to merge nodes, then fill the previous node to half
+      Return whether it's merged
+    */
+    pragma "no doc"
+    proc _merge(p: unmanaged _linkedNode(eltType)): bool {
+      var result = false;
+
+      // Nothing to do
+      if p.next == nil then return result;
+
+      var next = p.next!;
+
+      if p.size < nodeCapacity/2 {
+        // Fill p to half
+        var toFill = nodeCapacity/2 - p.size;
+        if next.size - toFill < nodeCapacity/2 {
```

Comment:
my reflex is to create a `const halfCap = nodeCapacity/2` and use `halfCap` to avoid repeated divisions here. But that is probably a premature optimizaiton

Reply:
I think the compiler should take care of that. So in the term of cleanness, which do you prefer? I'm on the fence.

Reply:
It probably doesn't today unless the backend is able factor it out, and I don't have strong opinions as this function probably have other parts that consume more time. We can leave it as-is.